### ISSUE TITLE
twisted/web/proxy.py doesn't use IRequest.finish

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2677]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2677|
|Type|defect|
|Created|2007-05-24 23:28:07Z|
|Branch|https://github.com/twisted/twisted/tree/finish-proxied-requests-2677|

```ProxyClient.handleResponseEnd``` uses ```loseConnection``` to disconnect the tcp session.  It should be calling finish on the request, though.

<details><summary>Searchable metadata</summary>

```
trac-id__2677 2677
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_finish_proxied_requests_2677 branches/finish-proxied-requests-2677
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__tape tape
time__1180049287000000 1180049287000000
changetime__1240321484000000 1240321484000000
version__None None
owner__ 
cc__jack
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|jack commented|
|-|-|

Specifically this manifested for me as ReverseProxyResource requests not appearing in the logs.  I have not found a work around.

As reverse proxies are becoming extremely common due to JavaScript's same origin policy, perhaps the priority of this bug should be stepped up a little?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Given how infrequently sponsors have actually requested things, I'm giving this a bump :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [26716]) Branching to 'finish-proxied-requests-2677'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner|
|-|-|

Done, enjoyably.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/finish-proxied-requests-2677)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

 * There are some problems with 2.3, as you use sorted.
 * The `_finished` flag isn't tested, apparently.
 * Is the loseConnection call in handleResponseEnd still necessary? I suspect request.finish will do the right thing.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [26723]) Python 2.3 compat

refs [#2677](https://github.com/twisted/twisted/issues/2677)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [26724]) Add a test case which covers the duplicate protection in handleResponseEnd

refs [#2677](https://github.com/twisted/twisted/issues/2677)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [26725]) Add a test for the loseConnection call

refs [#2677](https://github.com/twisted/twisted/issues/2677)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

Thanks for the review!

  1. Replaced `sorted` with `list.sort` in r26723
  1. Added a test which covers `_finished` in r26724
  1. Added a test which requires the `loseConnection` call in order to pass in r26725 (it is necessary because it's the outgoing request's transport, whereas the `request.finish()` call will only deal with the incoming request's transport).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set owner to @exarkun|
|-|-|

Great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [26738]) Fix the error case test to assert something valid; fix the implementation to use the right request methods to report the error

refs [#2677](https://github.com/twisted/twisted/issues/2677)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun removed owner|
|-|-|

I made a mistake in one of the tests.  It was asserting something nonsensical.  Fixed in the above linked revision.  Sorry about that.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set owner to @exarkun|
|-|-|

Fair enough, +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [26744]) Merge finish-proxied-requests-2677

Author: exarkun
Reviewer: therve
Fixes: [#2677](https://github.com/twisted/twisted/issues/2677)

Use the incoming `IRequest` object in the HTTP proxy when generating
a response to it rather than using its undocumented transport object.
This makes the proxy more generally useful (as it might support custom
request objects now), fixes a bug which led to proxied requests not
being logged, and simplifies the proxy code slightly be removing the
need for it to know much about the protocol-level formatting of HTTP.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
